### PR TITLE
Support all negative values in ClassLabel

### PR DESCRIPTION
**Pull request by lhoestq**
We usually use -1 to represent a missing label, but we should also support any negative values (some users use -100 for example). This is a regression from `datasets` 2.3

Fix https://github.com/huggingface/datasets/issues/4508

### CODE REVIEW

File: src/datasets/features/features.py [hunk 1]
```diff
@@ -1014,16 +1019,12 @@ def cast_storage(self, storage: Union[pa.StringArray, pa.IntegerArray]) -> pa.In
...
-                raise ValueError(f"Class label {min_max['min']} less than -1")
             if min_max["max"] >= self.num_classes:
                 raise ValueError(
                     f"Class label {min_max['max']} greater than configured num_classes {self.num_classes}"
                 )
         elif isinstance(storage, pa.StringArray):
-            storage = pa.array(
-                [self._strval2int(label) if label is not None else None for label in storage.to_pylist()]
-            )
+            storage = pa.array(self.str2int(storage.to_pylist()))
```
Comment (lhoestq) [inline src/datasets/features/features.py:1027]: Should we support `None` for the string representation of missing labels ? I think it could lead to silent errors in users code if we allow this

Comment (mariosasko) [reply src/datasets/features/features.py:1027]: Yes, I think we should still support this (to be consistent with the behavior during generation - `encode_example` is not called if a label is `None`)

Comment (lhoestq) [reply src/datasets/features/features.py:1027]: Ok ! `None` values can be used for N/A for example (which has a subtle difference compared to "unknown/missing label" represented by negative values).

Though to avoid silent error I would keep the current behavior of str2int and int2str, that is: raise an error if None is given as input

Comment (lhoestq) [reply src/datasets/features/features.py]: I reverted this to keep supporting None here :)

**Comment by davidbenton** [comment]
Thanks for this fix! I'm not sure what the release timeline is, but FYI #4508 is a breaking issue for transformer token classification using Trainer and PyTorch. PyTorch defaults to -100 as the ignored label for [log loss](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html?highlight=nllloss#torch.nn.NLLLoss), so switching labels to -1 leads to index errors using Trainer defaults.

As a workaround, I'm using master branch directly (`pip install git+https://github.com/huggingface/datasets.git@master` for anyone who needs to do the same) until this gets released.

**Comment by lhoestq** [comment]
The new release `2.4` fixes the issue, feel free to update `datasets` :) 
```
pip install -U datasets
```

**Comment by l-uuz** [comment]
@lhoestq I hope it's OK to ping you here. I've noticed that `encode_example` does only work with -1. I already created #7645 to fix the documentation, but then I stumbled across your original changes to the docs text in this PR.

I am talking about this part in `ClassLabel -> encode_example`:

https://github.com/huggingface/datasets/blob/e71b0b19d79c7531f9b9bea7c09916b5f6157f42/src/datasets/features/features.py#L1126-L1129
